platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/VariantIdentityUniquenessVerifier.java

```diff
@@ -0,0 +1,178 @@
+     * Build a report of all possible variant uniqueness failures for the given configurations.
+     */
+    public static VerificationReport buildReport(ConfigurationsProvider configurations) {
+        ListMultimap<VariantIdentity, ConfigurationInternal> byIdentity =
+            MultimapBuilder.linkedHashKeys().arrayListValues().build();
+
+        configurations.visitAll(configuration -> {
+            if (!mustHaveUniqueAttributes(configuration)) {
+                return;
+            }
+
+            byIdentity.put(VariantIdentity.from(configuration), configuration);
+        });
+
+        return new VerificationReport(byIdentity);
+    }
+
+    /**
+     * Consumable, non-resolvable, non-default configurations with attributes must have unique attributes.
+     */
+    private static boolean mustHaveUniqueAttributes(Configuration configuration) {
+        return configuration.isCanBeConsumed() &&
+            !configuration.isCanBeResolved() &&
+            !Dependency.DEFAULT_CONFIGURATION.equals(configuration.getName()) &&
+            !configuration.getAttributes().isEmpty();
+    }
+
+    /**
+     * A report tracking all possible variant uniqueness failures for a component.
+     */
+    public static class VerificationReport {
+
+        private final ListMultimap<VariantIdentity, ConfigurationInternal> byIdentity;
+
+        private VerificationReport(ListMultimap<VariantIdentity, ConfigurationInternal> byIdentity) {
+            this.byIdentity = byIdentity;
+        }
+
+        /**
+         * Get a failure that only checks variant uniqueness for the given configuration.
+         */
+        @Nullable
+        public GradleException failureFor(ConfigurationInternal configuration, boolean withTaskAdvice) {
+            List<ConfigurationInternal> collisions =
+                byIdentity.get(VariantIdentity.from(configuration)).stream()
+                    .filter(it -> it != configuration)
+                    .collect(Collectors.toList());
+
+            if (collisions.isEmpty()) {
+                return null;
+            }
+
+            DocumentedFailure.Builder builder = DocumentedFailure.builder();
+            String advice = "Consider adding an additional attribute to one of the configurations to disambiguate them.";
+            if (withTaskAdvice) {
+                advice += "  Run the 'outgoingVariants' task for more details.";
+            }
+
+            String message = "Consumable configurations with identical capabilities within a project (other than the default configuration) " +
+                "must have unique attributes, but " + configuration.getDisplayName() + " and " + collisions + " contain identical attribute sets.";
+
+            return builder.withSummary(message)
+                .withAdvice(advice)
+                .withUserManual("upgrading_version_7", "unique_attribute_sets")
+                .build();
+        }
+
+        /**
+         * Get a failure that encompasses all configurations.
```

Comment:
❗  This only gets the first failure it encounters during iteration though.  If there are multiple different duplicate identities they won't be returned.  

Like, if configuration A == B, and C == D, then you have a multimap of (identity 1 -> A, B), (identity 2 -> C, D) and you only get A, B out of this.

Was this meant to show all collisions?

Reply:
It was meant to do the same thing that the previous code did. The comment is misleading though. I've reworded it so that it no longer implies the exception encompasses all conflicts.

Reply:
That's not to say that this message can't be improved, I just did not intend to spend the time improving it with this change